js_utils/protocol_validation.js

```diff
@@ -0,0 +1,236 @@
+  const value_in_millis = +value_str * TIME_CONVERSION_TO_MILLIS[selected_unit];
+  let error_message;
+
+  if (value_str === "") {
+    error_message = invalid_err_msg.required;
+  } else if (isNaN(+value_str)) {
+    error_message = invalid_err_msg.delay_num_err;
+  } else if (value_in_millis < MIN_SUBPROTOCOL_DURATION_MS) {
+    error_message = invalid_err_msg.min_delay_duration;
+  } else if (value_in_millis > MAX_SUBPROTOCOL_DURATION_MS) {
+    error_message = invalid_err_msg.max_delay_duration;
+  } else if (!Number.isInteger(value_in_millis)) {
+    error_message = invalid_err_msg.non_integer;
+  } else {
+    error_message = invalid_err_msg.valid;
+  }
+
+  return error_message;
+};
+
+export const get_max_pulse_duration_for_freq = (freq) => {
+  return Math.min(50, Math.trunc((1000 / freq) * 0.8));
+};
+
+export const get_total_active_duration = (type, protocol) => {
+  return type === "Monophasic"
+    ? +protocol.phase_one_duration
+    : +protocol.phase_one_duration + +protocol.phase_two_duration + +protocol.interphase_interval;
+};
+
+export const is_valid_single_pulse = (protocol) => {
+  const { duration, unit } = protocol.total_active_duration;
+  const is_monophasic = protocol.type === "Monophasic";
+  const charges_to_check = is_monophasic ? ["phase_one_charge"] : ["phase_one_charge", "phase_two_charge"];
+  const durations_to_check = is_monophasic
+    ? ["phase_one_duration"]
+    : ["phase_one_duration", "phase_two_duration", "interphase_interval"];
+
+  const max_pulse_duration_for_freq = get_max_pulse_duration_for_freq(protocol.frequency);
+  const total_active_duration = get_total_active_duration(protocol.type, protocol);
+
+  // first check all durations are within max and min bounds
+  const durations_are_valid =
+    durations_to_check.filter(
+      (duration) =>
+        check_pulse_duration_validity(
+          protocol[duration],
+          duration === "interphase_interval",
+          max_pulse_duration_for_freq,
+          total_active_duration
+        ) !== ""
+    ).length === 0;
+
+  // check if charges are within max and min bounds
+  const charges_are_valid =
+    charges_to_check.filter((charge) => check_pulse_charge_validity(protocol[charge]) !== "").length === 0;
+
+  const complete_pulse_validity =
+    check_pulse_frequency_validity(protocol.frequency, max_pulse_duration_for_freq) === "" &&
+    check_active_duration_validity(duration, unit, total_active_duration) === "" &&
+    check_num_cycles_validity(protocol.num_cycles) === "";
+
+  return durations_are_valid && charges_are_valid && complete_pulse_validity;
+};
+
+export const is_valid_delay_pulse = (protocol) => {
+  const { duration, unit } = protocol;
+  return check_delay_pulse_validity(duration, unit) === "";
+};
+
+/*
+  BIPHASIC
+  ^^^^^^^^
+  frequency
+  interphase_interval
+  num_cycles
+  phase_one_charge
+  phase_one_duration
+  phase_two_charge
+  phase_two_duration
+  postphase_interval
+  total_active_duration
```

Comment:
Did we add handling for importing protocols with the old format? I thought we got rid of this field